File: cpp/leetcode/439.cpp
class Solution {
public:
    string parseTernary(string& expression, int begin = 0) {
        if (begin >= expression.size()) return "";
        if (begin == expression.size() - 1 || expression[begin + 1] == ':') return expression.substr(begin, 1);
        if (expression[begin] == 'T') return parseTernary(expression, begin + 2);
        int level = 1, i = begin + 2;
        for (; i < expression.size() && level; i++) {
            if (expression[i] == '?') level++;
            else if (expression[i] == ':') level--;
        }
        return parseTernary(expression, i);
    }
};

File: python/leetcode/439.py
class Solution(object):
    def parseTernary(self, expression):
        while len(expression) != 1:
            i = expression.rindex("?")    # begin with the last '?'.
            tmp = expression[i+1] if expression[i-1] == 'T' else expression[i+3]
            expression = expression[:i-1] + tmp + expression[i+4:]
        return expression